---
title: "Flexible iterative fitting over omics-type data using the modelling algorithm of your choice"
output: rmarkdown::html_vignette
bibliography: references.bib
author: "Daniel HammarstrÃ¶m"
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{Flexible iterative fitting over omics-type data using the modelling algorithm of your choice}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



## Background

Differential expression is determined in RNA-sequencing data using specific software packages such as edgeR [@McCarthy2012] or DEseq2 [@Love2014]. These packages provides efficient algorithms for fitting gene-expression data to count-based data (i.e. RNA-seq data) by sharing information across genes in relatively simple designs. More elaborate experimental designs may require the addition of random effects to such models when observations are not independent. To accommodate random effects Vestal et al. [@Vestal2022] developed [lmerSeq](https://github.com/stop-pre16/lmerSeq) which allowed iterative fitting of mixed linear models to transformed count data. Similarly, Lewis et al. [@Lewis] developed [glmmSeq](https://cran.r-project.org/web/packages/glmmSeq/vignettes/glmmSeq.html#glmmSeq) to fit negative binomial mixed effects models. We have previously used negative binomial mixed models to data from the trainome-project [@Khan2020; @Mlmen2021], however the software implementation was not formalized to user-friendly functions.

The aim of the `seq_wrapper` function is to provide even more flexibility to model fitting of omics-type data. Although [lmerSeq](https://github.com/stop-pre16/lmerSeq) and [glmmSeq](https://cran.r-project.org/web/packages/glmmSeq/vignettes/glmmSeq.html#glmmSeq) both provides powerful implementations of fitting algorithms to iterative fitting they are limited as the user cannot select which fitting algorithm to use. `seq_wrapper` can use any available fitting algorithm (stats::lm, glmmTMB::glmmTMB, lme4::lmer, etc.) which makes the model fitting process more flexible. In addition, the user can provide user-created summary- and model evaluation-functions to tailor output for specific purposes.

## Iterative fitting of high dimensional data

The `seq_wrapper` function lets the user specify which modelling algorithm to be used in modelling data provided as a data frame of counts (or similar) and metadata. Target-level models are passed to the results together with optional user-specified summary and model-diagnostic functions. The function aims to not limit the user in preferences of specific models. This allows for comparisons between modelling frameworks and flexibility for many types of omics-type data. The basic outline of the function can be seen in the diagram below.


```{r diagram}
#| echo: false
#| fig-cap: "Basic structure of the `seq_wrapper` function. Data and metadata together with user-selected modelling algorithms and arguments are combined in parallel to fit model objects. Results are returned containing models and any results from user-constructed summaries or diagnostics." 
library(DiagrammeR)

DiagrammeR::grViz("             # All instructions are within a large character string
digraph surveillance_diagram {  # 'digraph' means 'directional graph', then the graph name 
  
  # graph statement
  #################
  forcelabels=true;
  graph [layout = dot,
         rankdir = TB,            
         overlap = false,
         fontsize = 8]
  

  # nodes (circles)
  #################
  node [shape = circle,                  # shape = circle
       fixedsize = true
       width = 1.3]                      # width of circles
  

    data   [label = 'High-dimension\ndata'] 
    metadata [label = 'Meta data'] 
    model [label = 'Modelling\nalgorithm'] 
    modelargs [label = 'Model\narguments'] 
    
    fit [label = 'Model\nobjects'] 
    summaries [label = 'Model\nsummaries'] 
    evaluations [label = 'Model\ndiagnostics'] 

    output [label = 'Results']

  
  # edges
  #######

  data -> fit
  metadata -> fit
  model -> fit
  modelargs -> fit
  
  fit -> summaries
  fit -> evaluations
  
  fit -> output
  summaries -> output
  evaluations -> output


}
")





```






```{r packages}
 library(trainomeHelper); library(tidyverse)
 
 # load sample data 
 data(rna_seq_sample); data(rna_seq_metadata)
 
 
 ## Check the number of cores available
 parallel::detectCores()
 
 
 # 
 # 
  fits <- seq_wrapper(    fitting_fun = glmmTMB::glmmTMB,
                          arguments = list(formula = counts ~ time + time:condition + (1|participant), 
                                           family = glmmTMB::nbinom2),
                          data = rna_seq_sample,
                          metadata = rna_seq_metadata,
                          samplename = "seq_sample_id",
                          summary_fun = NULL,
                          eval_fun = NULL,
                          additional_vars = NULL,
                          exported = list(),
                          subset = NULL,
                          cores = 8)
  
  

```

### Summary functions

`seq_wrapper` accepts user-defined summary functions.

```{r}

x <- fits$model_fits[[1]]



sum_fun <- function(x){
  
  cond_effects <- data.frame(cbind(data.frame(coef = rownames(coef(summary(x))$cond))),
                             coef(summary(x))$cond, 
                                    
                             row.names = NULL)
  
  return(cond_effects)
  
}



  fits <- seq_wrapper(    fitting_fun = lme4::lmer,
                          arguments = list(formula = counts ~ time + time:condition + (1|participant)),
                          data = rna_seq_sample,
                          metadata = rna_seq_metadata,
                          samplename = "seq_sample_id",
                          summary_fun = NULL,
                          eval_fun = NULL,
                          additional_vars = NULL,
                          exported = list(),
                          subset = c(1,34, 67,8,9),
                          cores = 1)

c(1,34, 67,8,9)

  bind_rows(fits$model_summarises) %>%
    mutate(target = rep(names(fits$model_summarises), each = 8)) %>%
    print()
  
  

sumaries <- lapply(fits$model_fits, FUN = sum_fun) 

bind_rows(sumaries) %>%
  mutate(target = rep(names(fits$model_fits), each = 8)) %>%
  group_by(coef) %>%
  mutate(fdr = p.adjust(Pr...z.., method = "fdr")) %>%
  
  filter(coef == "timew2pre:conditionsingle") %>%
  arrange(fdr)
   

  
  
  ggplot(aes(Pr...z..)) + geom_histogram() + facet_wrap(~ coef)

```








## References



